From: Li Yu <bingtian.ly@taobao.com>
Subject: [PATCH] igb: fix default behavior of RSS parameter to automatically detect number of Rx/Tx queues
Patch-mainline: in-house-fix

The code comments in igb_param.c is wrong, it only bring up one Tx/Rx queue,
which should be "default 0=number of cpus". By the fix, if no parameter is
specified by RSS, default detected Tx/Rx queues number is minimal number
between opt.arg.r.max and num_online_cpus.

Signed-off-by: Li Yu <bingtian.ly@taobao.com>
Reviewed-by: Coly Li <bosong.ly@taobao.com>
---
Index: linux-2.6.32-220.2.1.el5/drivers/net/igb/igb_param.c
===================================================================
--- linux-2.6.32-220.2.1.el5.orig/drivers/net/igb/igb_param.c	2012-02-15 20:11:30.979609467 +0800
+++ linux-2.6.32-220.2.1.el5/drivers/net/igb/igb_param.c	2012-02-15 20:23:41.835233596 +0800
@@ -132,11 +132,11 @@
  *
  * Valid Range: 0 - 8
  *
- * Default Value:  1
+ * Default Value:  0
  */
-IGB_PARAM(RSS, "Number of Receive-Side Scaling Descriptor Queues (0-8), default 1=number of cpus");
+IGB_PARAM(RSS, "Number of Receive-Side Scaling Descriptor Queues (0-8), default 0=number of cpus");
 
-#define DEFAULT_RSS       1
+#define DEFAULT_RSS       0
 #define MAX_RSS           ((adapter->hw.mac.type == e1000_82575) ? 4 : 8)
 #define MIN_RSS           0
 
@@ -575,7 +575,8 @@
 			}
 #ifdef module_param_array
 		} else {
-			adapter->rss_queues = opt.def;
+			adapter->rss_queues = opt.def ?:
+				min_t(u32, opt.arg.r.max, num_online_cpus());
 		}
 #endif
 	}
